upgrades/steps_28.go

```diff
@@ -321,6 +325,137 @@ func readStorageStateFile(dirPath string, tag names.StorageTag) (bool, error) {
 	return *info.Attached, nil
 }
 
+// storageDiskInfo defines the storage unit storageData serialization.
 type storageDiskInfo struct {
 	Attached *bool `yaml:"attached,omitempty"`
 }
+
+type relationData struct {
+	data     map[int]string
+	filename string
+}
+
+func readRelationState(uniterStateDir string) (relationData, error) {
+	relationStateDir := filepath.Join(uniterStateDir, "relations")
+
+	rSt, err := readAllRelationStateDirs(relationStateDir)
+	if err != nil {
+		return relationData{}, err
+	}
+	retData := relationData{data: make(map[int]string, len(rSt)), filename: relationStateDir}
+	for k, v := range rSt {
+		dataYaml, err := yaml.Marshal(v)
+		if err != nil {
+			return relationData{}, errors.Annotatef(err, "failed to unmarshal relation state from %q", relationStateDir)
+		}
+		retData.data[k] = string(dataYaml)
+	}
+	return retData, nil
+}
+
+// readAllRelationStateDirs loads and returns every relationState persisted directly inside
+// the supplied dirPath. If dirPath does not exist, no error is returned.
+func readAllRelationStateDirs(dirPath string) (dirs map[int]*relationState, err error) {
+	defer errors.DeferredAnnotatef(&err, "cannot load relations state from %q", dirPath)
+	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	fis, err := ioutil.ReadDir(dirPath)
+	if err != nil {
+		return nil, err
+	}
+	dirs = map[int]*relationState{}
+	for _, fi := range fis {
+		// Entries with integer names must be directories containing StateDir
+		// storageData; all other names will be ignored.
+		relationId, err := strconv.Atoi(fi.Name())
+		if err != nil {
+			// This doesn't look like a relation.
+			continue
+		}
+		rState, err := readRelationStateDir(filepath.Join(dirPath, fi.Name()), relationId)
+		if err != nil {
+			return nil, err
+		}
+		dirs[relationId] = rState
+	}
+	return dirs, nil
+}
+
+// readRelationStateDir loads a relationState from the subdirectory of dirPath named
+// for the supplied RelationId. If the directory does not exist, no error
+// is returned,
+func readRelationStateDir(dirPath string, relationId int) (d *relationState, err error) {
+	d = &relationState{
+		RelationId:         relationId,
+		Members:            map[string]int64{},
+		ApplicationMembers: map[string]int64{},
+		ChangedPending:     "",
+	}
+	defer errors.DeferredAnnotatef(&err, "cannot load relation state from %q", dirPath)
+	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
+		return d, nil
```

Comment:
Same question about suppressing the error and returning back a blank relationState. So, if we find an _empty_ folder for the relation we silently convert it into a blank state? Wouldn't the absense of state data indicate that something is wrong?